Correctness/Correctness.lean

```diff
@@ -175,4 +215,431 @@ theorem partial_correctness_from_verification_conditions [Sys σ] [Spec' σ]
+                k ≤ n ∧ cut (run si k) :=
+     if hn : i < n then
+       if hc : cut s' then
+         have := cassert_cut hc i
+         have h_i_i : i - i = 0 := by simp only [Nat.sub_self]
+         ⟨i, this.left, h_i_i ▸ this.right, hle, by subst s'; assumption⟩
+       else
+         have ⟨k, hs, hkn, hck⟩ := loop (next s') (i+1) hn (by simp [heq, run_succ'])
+         have h_not_cut := (cassert_not_cut hc hs).right
+         have : (cassert s0 s' i).snd = assert s0 (run s' (k - i)) := by
+          simp only [h_not_cut, hkn]
+          have h_next_cut : cut (run (next s') (n - i - 1)) := by
+            rw [heq]
+            have : (next (run si i)) = run (run si i) 1 := by
+              simp only [run]
+            simp only [this]
+            have : (run (run (run si i) 1) (n - i - 1)) = run si n := by
+              repeat (rw [run_run])
+              have : (i + (1 + (n - i - 1))) = n := by omega
+              rw [this]
+              done
+            simpa only [this]
+          have h_lb := @cassert_lower_bound _ (i + 1) k _ _ s0
+                        (next s') (n - i - 1) h_next_cut hs
+          conv =>
+            rhs
+            unfold run
+          have h_k_i : ¬ k - i = 0 := by omega
+          split
+          · contradiction
+          · rename_i x y h_k_i_y
+            simp only [Nat.succ_eq_add_one] at h_k_i_y
+            have : (k - (i + 1)) = k - i - 1 := by omega
+            simp only [this, h_k_i_y, Nat.add_one_sub_one]
+          done
+         ⟨k, (cassert_not_cut hc hs).left, this, hck.left, hck.right⟩
+     else
+       have hin : i = n := by omega
+       have h_left : (cassert s0 s' i).fst = n := by
+          have := @cassert_cut _ _ _ s0 _ hc n
+          subst hin
+          rw [heq]
+          simp only [this]
+       have h_right : (cassert s0 s' i).snd = assert s0 (run s' (n - i)) := by
+        simp only [Nat.sub_self, eq_iff_iff]
+        have := @cassert_cut _ _ _ s0 _ hc
+        subst hin
+        rw [heq]
+        simp only [this, Nat.sub_self, eq_iff_iff, run]
+        done
+       ⟨n, h_left, h_right, Nat.le_refl .., by assumption⟩
+  loop si 0 (Nat.zero_le ..) rfl
+
+/--
+Prove partial correctness from inductive assertions using `cassert`
+function.
+
+We use `s0`, `si`, and `sf` to refer to initial, intermediate, and
+final (exit) states respectively.
+
+This is more convenient to use that
+`partial_correctness_from_inductive_assertions` because we can do
+symbolic simulation and open `cassert` in tandem.
+-/
+theorem partial_correctness_from_assertions [Sys σ] [Spec' σ]
+    (v1 : ∀ s0 : σ, pre s0 → assert s0 s0)
+    -- (FIXME) Is it possible to remove v2 and combine v3 and v4 into
+    -- a single verification condition as follows?
+    -- As @bollu noted, this has the benefit of not having `post` be a
+    -- callee of `assert`.
+    --
+    --  ∀ s0 si : σ, assert s0 si → (cassert s0 (run si 1) 0).snd ∨
+    --                              (exit si → post s0 si)
```

Comment:
Shouldn't it be `exit si /\ post s0 si`? Otherwise, the condition will be trivial for non-exit states

Reply:
Yes, thanks for noticing. What's written is definitely wrong, and I haven't figured out the right form yet (hence the TODO) :) 
Let's take up what you suggest:
`a -> c or (e and p)` (I'm using `a` for `assert`, `c` for cassert`, `e` for `exit`, and `p` for `post`).

Suppose `e` is `true`. In that case, we have:
`a -> c or p`

But we really want `p` to hold at an exit state, and this formulation too does not enforce that.

We can combine `v3` and `v4` to be something like:
`a -> (e or c) and (not e or p)`
but I need to work out how to remove `v2`.